src/roots.rs

```diff
@@ -5,17 +5,104 @@ use std::{
+/// under the specific root is included.
+///
+/// *NOTE*: If the parent folder of a file is not included, then
+/// `include_file` will not be called.
+///
+/// # Example
+///
+/// Implementing `Filter` for rust files:
+///
+/// ```
+/// use ra_vfs::{Filter, RelativePath};
+///
+/// struct IncludeRustFiles;
+///
+/// impl Filter for IncludeRustFiles {
+///     fn include_folder(&self, folder_path: &RelativePath) -> bool {
+///         // These folders are ignored
+///         const IGNORED_FOLDERS: &[&str] = &["node_modules", "target", ".git"];
+///
+///         let is_ignored = folder_path.components().any(|c| IGNORED_FOLDERS.contains(&c.as_str()));
+///
+///         !is_ignored
+///     }
+///
+///     fn include_file(&self, file_path: &RelativePath) -> bool {
+///         // Only include rust files
+///         file_path.extension() == Some("rs")
+///     }
+/// }
+/// ```
+pub trait Filter: Send + Sync {
+    fn include_folder(&self, folder_path: &RelativePath) -> bool;
+    fn include_file(&self, file_path: &RelativePath) -> bool;
+}
+
+/// RootEntry identifies a root folder with a given filter
+/// used to determine whether to include or exclude files and folders under it.
+pub struct RootEntry {
+    path: PathBuf,
+    filter: Box<dyn Filter>,
+}
+
+impl std::fmt::Debug for RootEntry {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "RootEntry({})", self.path.display())
+    }
+}
+
+impl Eq for RootEntry {}
+impl PartialEq for RootEntry {
+    fn eq(&self, other: &Self) -> bool {
+        // Entries are equal based on their paths
+        self.path == other.path
+    }
+}
+
+impl RootEntry {
+    /// Create a new `RootEntry` with the given `filter` applied to
+    /// files and folder under it.
+    pub fn new(path: PathBuf, filter: Box<dyn Filter>) -> Self {
+        RootEntry { path, filter }
+    }
+}
+
 /// VfsRoot identifies a watched directory on the file system.
 #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct VfsRoot(pub u32);
 
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(crate) enum FileType {
+    File,
+    Dir,
+}
+
+impl FileType {
+    pub(crate) fn is_dir(&self) -> bool {
+        *self == FileType::Dir
+    }
+}
+
+impl std::convert::From<std::fs::FileType> for FileType {
+    fn from(v: std::fs::FileType) -> Self {
+        if v.is_file() {
+            FileType::File
+        } else {
+            FileType::Dir
+        }
+    }
+}
+
 /// Describes the contents of a single source root.
 ///
-/// `RootConfig` can be thought of as a glob pattern like `src/**.rs` which
+/// `RootData` can be thought of as a glob pattern like `src/**.rs` which
 /// specifies the source root or as a function which takes a `PathBuf` and
-/// returns `true` iff path belongs to the source root
+/// returns `true` if path belongs to the source root
 struct RootData {
-    path: PathBuf,
+    entry: RootEntry,
```

Comment:
I think it'd be more clear if we remove `RootEntry` here and keep `root` and `filter`